include/event_manager/event_manager.h

```diff
@@ -0,0 +1,317 @@
+#define ZEPHYR_INCLUDE_EVENT_MANAGER_EVENT_MANAGER_H_
+
+
+/**
+ * @defgroup event_manager Event Manager
+ * @brief Event Manager
+ *
+ * @{
+ */
+
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <sys/reboot.h>
+#include <sys/__assert.h>
+#include <logging/log_ctrl.h>
+
+#include <event_manager/event_manager_priv.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/** @def SUBS_PRIO_MIN
+ *
+ * @brief Index of the highest subscriber priority level.
+ */
+#define SUBS_PRIO_MIN    _SUBS_PRIO_FIRST
+
+
+/** @def SUBS_PRIO_MAX
+ *
+ * @brief Index of the lowest subscriber priority level.
+ */
+#define SUBS_PRIO_MAX    _SUBS_PRIO_FINAL
+
+
+/** @def SUBS_PRIO_COUNT
+ *
+ * @brief Number of subscriber priority levels.
+ */
+#define SUBS_PRIO_COUNT (SUBS_PRIO_MAX - SUBS_PRIO_MIN + 1)
+
+
+/** @brief Event header.
+ *
+ * When defining an event structure, the event header
+ * must be placed as the first field.
+ */
+struct event_header {
+	/** Linked list node used to chain events. */
+	sys_snode_t node;
+
+	/** Pointer to the event type object. */
+	const struct event_type *type_id;
+};
+
+
+/** @brief Dynamic event data.
+ *
+ * When defining an event structure, the dynamic event data
+ * must be placed as the last field.
+ */
+struct event_dyndata {
+	/** Size of the dynamic data. */
+	size_t size;
+
+	/** Dynamic data. */
+	uint8_t data[0];
+};
+
+
+/** @brief Event listener.
+ *
+ * All event listeners must be defined using @ref EVENT_LISTENER.
+ */
+struct event_listener {
+	/** Name of this listener. */
+	const char *name;
+
+	/** Pointer to the function that is called when an event
+	 *  is handled.
+	 */
+	bool (*notification)(const struct event_header *eh);
+};
+
+
+/** @brief Event subscriber.
+ */
+struct event_subscriber {
+	/** Pointer to the listener. */
+	const struct event_listener *listener;
+};
+
+
+/** @brief Event type.
+ */
+struct event_type {
+	/** Event name. */
+	const char			*name;
+
+	/** Array of pointers to the array of subscribers. */
+	const struct event_subscriber	*subs_start[SUBS_PRIO_COUNT];
+
+	/** Array of pointers to the element directly after the array of
+	 * subscribers.
+	 */
+	const struct event_subscriber	*subs_stop[SUBS_PRIO_COUNT];
+
+	/** Bool indicating if the event is logged by default. */
+	bool init_log_enable;
+
+	/** Function to log data from this event. */
+	int (*log_event)(const struct event_header *eh, char *buf,
+			      size_t buf_len);
+
+	/** Custom data related to tracking. */
+	const void *trace_data;
+};
+
+
+extern const struct event_listener __start_event_listeners[];
+extern const struct event_listener __stop_event_listeners[];
+
+extern const struct event_type __start_event_types[];
+extern const struct event_type __stop_event_types[];
+
+
+/** Create an event listener object.
+ *
+ * @param lname   Module name.
+ * @param cb_fn  Pointer to the event handler function.
+ */
+#define EVENT_LISTENER(lname, cb_fn) _EVENT_LISTENER(lname, cb_fn)
```

Comment:
internal api's should be prefixed with `Z_`/`z_` instead of `_`

Reply:
@nordic-krch I am thinking how to approach it. So far we kept the changes in sync with NCS. This is needed as we don't have manpower to maintain this PR for long period of time while still maintaining NCS version. I don't think we can apply this change in NCS.
Maybe the renames can be done as a last step when we get to the point PR is accepted?

Reply:
sure.